src/platforms/stm32/src/lib/gpio_driver.h

```diff
@@ -26,13 +26,62 @@
+static const char *const rising_atom = ATOM_STR("\x6", "rising");
+static const char *const falling_atom = ATOM_STR("\x7", "falling");
+static const char *const both_atom = ATOM_STR("\x4", "both");
+static const char *const set_level_atom = ATOM_STR("\x9", "set_level");
+static const char *const read_atom = ATOM_STR("\x4", "read");
+static const char *const set_direction_atom = ATOM_STR("\xD", "set_direction");
+static const char *const set_int_atom = ATOM_STR("\x7", "set_int");
+static const char *const remove_int_atom = ATOM_STR("\xA", "remove_int");
+static const char *const close_atom = ATOM_STR("\x5", "close");
+static const char *const a_atom = ATOM_STR("\x1", "a");
+static const char *const b_atom = ATOM_STR("\x1", "b");
+static const char *const c_atom = ATOM_STR("\x1", "c");
+static const char *const d_atom = ATOM_STR("\x1", "d");
+static const char *const e_atom = ATOM_STR("\x1", "e");
+static const char *const f_atom = ATOM_STR("\x1", "f");
+static const char *const g_atom = ATOM_STR("\x1", "g");
+static const char *const h_atom = ATOM_STR("\x1", "h");
+#ifdef LIBOPENCM3_GPIO_COMMON_F24_H
+static const char *const i_atom = ATOM_STR("\x1", "i");
+static const char *const j_atom = ATOM_STR("\x1", "j");
+static const char *const k_atom = ATOM_STR("\x1", "k");
+#endif /* defined LIBOPENCM3_GPIO_COMMON_F24_H */
+static const char *const mhz_2_atom = ATOM_STR("\x5", "mhz_2");
+static const char *const mhz_25_atom = ATOM_STR("\x6", "mhz_25");
+static const char *const mhz_50_atom = ATOM_STR("\x6", "mhz_50");
+static const char *const mhz_100_atom = ATOM_STR("\x7", "mhz_100");
+static const char *const low_atom = ATOM_STR("\x3", "low");
+static const char *const high_atom = ATOM_STR("\x4", "high");
+static const char *const up_atom = ATOM_STR("\x2", "up");
+static const char *const down_atom = ATOM_STR("\x4", "down");
+static const char *const floating_atom = ATOM_STR("\x8", "floating");
+static const char *const input_atom = ATOM_STR("\x5", "input");
+static const char *const output_atom = ATOM_STR("\x6", "output");
+static const char *const output_od_atom = ATOM_STR("\x9", "output_od");
+static const char *const af_atom = ATOM_STR("\x2", "af");
+static const char *const analog_atom = ATOM_STR("\x6", "analog");
+static const char *const invalid_bank_atom = ATOM_STR("\xC", "invalid_bank");
+static const char *const invalid_pin_atom = ATOM_STR("\xB", "invalid_pin");
+static const char *const invalid_mode_atom = ATOM_STR("\xC", "invalid_mode");
+static const char *const invalid_pull_atom = ATOM_STR("\xC", "invalid_pull");
+static const char *const invalid_rate_atom = ATOM_STR("\xC", "invalid_rate");
+static const char *const invalid_level_atom = ATOM_STR("\xD", "invalid_level");
+static const char *const invalid_trigger_atom = ATOM_STR("\xF", "invalid_trigger");
+static const char *const invalid_irq_atom = ATOM_STR("\xB", "invalid_irq");
+static const char *const gpio_interrupt_atom = ATOM_STR("\xE", "gpio_interrupt");
```

Comment:
The usual habit for defining variables in header files is using `extern` to avoid allocating them every time the .h file is included, and deciding an .c file that allocates them (and defines the value to which they will be initialized).

However,

regardless of what I wrote just before,


do we really have any advantage in defining all those variables in `gpio_driver.h`? I don't think so, lot of them aren't even used outside of GPIO driver (such as `set_direction`).

Reply:
I was that most of those will be used in the SPI driver. There is already a mechanism to include or not the gpio nifs and port driver, I plan on the same for SPI. For extremely constrained systems it is possible SPI will be desired, but gpio functionality will not be required. In that case gpio_driver.h could be included in the SPI driver, even if the port driver and nifs were excluded.

Reply:
Thank you for pointing that out, I was unaware of the extern pattern. In this case, in light of my anticipation of needing these atoms in the spi and I2C setup routines I think a gpio_atoms.c makes sense for initializing and it can be used by any driver that requires them.

Reply:
I moved these back into `gpio_driver.c` and reduced the number of `static const char *const` to the minimum necessary atoms.